Railsのテストを書く

YuSa0-6 opened the pull request:
# プラクティス「Railsでテストを書く」の提出
## なぜやったか
- report, userクラスのインスタンスにテストを書いてRailsでのテストの書き方を知るため
- fixturesの概要を知るため
- Report,Userモデルに対して期待通りに動くことを確認するため。
## 何をやったか
- Report、Userモデルのメゾットに対してテストを行う
　- private内のメゾットに対しては書いてないです。
- メルアド、パスワードでログイン後、日報のCRUD処理のテストを行う
## 困ったこと
- Report.rbのprivateメゾットのテストのやり方として困った。
　- レコードが登録されたかどうかをテストするのか、DB制約をクリアしているのかどう対応するかが迷っている。
　　[Rails の ActiveRecord モデルテストの書き方ガイドライン - passingloopの日記](https://passingloop.hatenadiary.org/entry/20111125/how_to_write_rails_activerecord_model_tests)
    >その他のメソッドをテストしないのは？
[ActiveRecord](http://d.hatena.ne.jp/keyword/ActiveRecord) モデルクラスに定義されるメソッドは，インクルードされる論理モデルモジュールからのみ呼び出されるべきです．よって，これらのメソッドは基本的にはプライベートメソッドとして定義されます．[ユニットテスト](http://d.hatena.ne.jp/keyword/%A5%E6%A5%CB%A5%C3%A5%C8%A5%C6%A5%B9%A5%C8)でプライベートメソッドもテストすべきかどうかは議論が分かれるところですが，私は「プライベートメソッドまで無理してまでテストする必要はない．[Ruby](http://d.hatena.ne.jp/keyword/Ruby) だからできるけど」という考えなので，テストしません．
- systemテストのログイン処理に対して`Devise::Helper`が使えなくて困っていた。Wardenを使用して解決したが、少し不安。
- visit xx_urlの記述で画面遷移ができていないと考えたので、直接`reports/1`のように遷移するようにした。
　- 自分で調べてみます。

YuSa0-6 committed:
ActiveRecord::NotNullViolation: RuntimeError: NOT NULL constraint failed解消するために、それぞれのcontroller_test内のテストを削除する。

YuSa0-6 committed:
test:　Userモデルの`name_or_email`の正常家テストを行うために、ユーザーに名前がない場合はemailを返すようにテストを実装した。

YuSa0-6 committed:
テスト　その日報が編集可能かどうかの返り値をテストするために、ユーザー、日報を作成し、trueを正常とするテストを書く。

YuSa0-6 committed:
test: `models/report.rb` #created_onがDateオブジェクトを返しているかを確認するために、日報を新規作成後正常系テストを行う。

YuSa0-6 committed:
test: DB制約に関するエラーを回避するために、今回使用しなかったfixturesのデータ定義文を削除、またはコメントアウトする。

YuSa0-6 committed:
test: テストで使用するユーザーデータのために、email, name ,encrypted_passwordを定義する
- encrypted_passwordはdeviseでのパスワードを使用するために利用する
https://www.rubydoc.info/gems/devise/Devise/Encryptor#digest-class_method

YuSa0-6 committed:
test: 今回のプラクティでBooksのSystemテストを行わないため、テストを削除する。

YuSa0-6 committed:
test: ログイン、日報のCURDの一連の流れをテストする。
- include Devise::Testhelpersが動かなかったため、Wardenのヘルパーを使用する。
- https://www.rubydoc.info/github/hassox/warden/Warden/Test/Helpers

yoshitsugu commented on test/models/report_test.rb:
```diff
@@ -6,4 +6,14 @@ class ReportTest < ActiveSupport::TestCase
   #   assert true
   # end
+  test '#editable?' do
+    user = User.new(email: 'test@sample.com', password: 'password')
+    report = user.reports.new(title: 'title', content: 'content')
```
注意点に

>  テストデータの作成はfixturesの使用を推奨する（どうしても使い方がわからない場合はfixturesなしで、Rubyのロジックを書いてテストデータを作成しても良い）

と書かれている通りfixturesを使うようにしてください。

yoshitsugu commented on test/models/report_test.rb:
```diff
@@ -6,4 +6,14 @@ class ReportTest < ActiveSupport::TestCase
+  test '#editable?' do
+    user = User.new(email: 'test@sample.com', password: 'password')
+    report = user.reports.new(title: 'title', content: 'content')
+    assert_equal true, report.editable?(user)
```
editable?がfalseになる場合もテストしましょう

yoshitsugu commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,9 @@
+  test '#name_or_email' do
+    user = User.new(email: 'test9@sample.com', name: '')
+
+    assert_equal 'test9@sample.com', user.name_or_email
```
nameになる場合とemailになる場合両方テストしましょう。

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -3,45 +3,37 @@
 
 class ReportsTest < ApplicationSystemTestCase
+  include Warden::Test::Helpers
+
   setup do
-    @report = reports(:one)
+    Warden.test_mode!
+    @user = users(:alice)
+    login_as(@user, scope: :user)
+    @report = @user.reports.create(title: '日報タイトル', content: '日報の内容')
```
ログインもSystemTestとしてログイン画面を操作してログインするようにしましょう。

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -3,45 +3,37 @@
+    fill_in 'タイトル', with: '新規日報'
+    fill_in '内容', with: '今日は晴天なり'
+    click_on '登録する'
+    assert_text '日報が作成されました。'
```
作成した内容が表示されていることまでテストしましょう

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -3,45 +3,37 @@
+    fill_in 'タイトル', with: '更新した日報'
+    fill_in '内容', with: '今日は雨'
+    click_on '更新する'
+    assert_text '日報が更新されました。'
```
更新した内容が表示されていることまでテストしましょう

YuSa0-6 replied:
-[ ] fill_inされた内容かどうかを確認する

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -3,45 +3,37 @@
+  test 'destroy a report' do
+    visit '/reports/1'
+    click_on 'この日報を削除'
+    assert_text '日報が削除されました。'
```
削除した内容が表示されなくなったことまでテストしましょう

yoshitsugu reviewed (commented):
書いておられる通り、 `save_mentions` 直接のテストは必要ありません。
ただ、このprivateメソッドの定義により `save` などの挙動の期待値が変わっていますよね。そこをテストするようにしましょう。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#discussion_r1586927100
テストデータを作成するために、fixturesを利用する。
- `users(:alice)にてusers.yml alice`のデータを取得する
- `reports.((:report_alice)).attributes`にて関連するreportのデータを取得する

YuSa0-6 committed:
[review url](https://github.com/YuSa0-6/fjord-books_app-2023/pull/8/files/c71ce64c19a4bd134dc6e7a0911deedc7a5eaa64#r1586927206)
editable?がfalseになる場合もテストをする。
そのためにnameがないusers(:guest)を用意する。

YuSa0-6 committed:
[review url](https://github.com/YuSa0-6/fjord-books_app-2023/pull/8/files/c71ce64c19a4bd134dc6e7a0911deedc7a5eaa64#r1586927359)
name_or_emailのnameになる場合とemailになる場合両方をテストする。
aliceにはname: 'alice'を設定する。
guestにはname: ''を設定する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#discussion_r1586928035
ログインもSystemテストとして扱う。
setupにて事前にログインしているため、はじめにログアウトする。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#discussion_r1586928199
作成した内容まで表示されることまでテストする。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#discussion_r1586928264
更新された内容が表示されるところまでテストする。

YuSa0-6 committed:
[review]削除した内容がなくなったことまでテストする。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#pullrequestreview-2034617632
日報のURIを含む日報作成時にメンションする側とメンションされる側で登録できているかをテストする。

YuSa0-6 committed:
typoを修正する。

yoshitsugu commented on test/models/report_test.rb:
```diff
@@ -6,4 +6,31 @@ class ReportTest < ActiveSupport::TestCase
+    report = Report.new(created_at: Time.zone.local(2021, 1, 1, 12, 0, 0))
+    assert_equal Date.new(2021, 1, 1), report.created_on
+  end
+  test '#save_mentions with mentioning_reports' do
+    report = reports(:report_alice)
+    created_report = Report.create!(user: users(:alice), title: 'title', content: "http://localhost:3000/reports/#{report.id}")
+    assert_equal report.id, created_report.mentioning_reports.first.id
+  end
+
+  test '#save_mentions with mentioned_reports' do
+    report = reports(:report_alice)
+    created_report = Report.create!(user: users(:alice), title: 'title', content: "http://localhost:3000/reports/#{report.id}")
+    assert_equal created_report.id, report.mentioned_reports.first.id
+  end
```
作成時だけではなく、更新時に本文の内容に応じて、Mentionが追加、維持、削除されることまで確認しましょう。

yoshitsugu replied:
不足しているテストケースの指摘が多いように思いましたので、テスト技法のプラクティスを思い出してもらって、過不足ないテストを考えてみてくださいね。

YuSa0-6 replied:
「維持」の言葉のニュアンスが難しかったので質問させてください。
Systemテストとして、ちゃんと言及欄として表示できているかもテストするべきででしょうか？

yoshitsugu replied:
いえ、System testまでは必要ありません。
ここで「維持」と使ったのは単に追加も削除ものされていないもの、そのままキープされているもの、という意味で使いました。

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -3,45 +3,54 @@
+    visit "/reports/#{report.id}"
+    click_on 'この日報を削除'
+    assert_text '日報が削除されました。'
+    assert_equal true, Report.where(id: report.id).empty?
```
system テストではブラウザの表示を主にテストします。バックエンドがどうなっているかはブラックボックスとして扱うことが多いです。今回もDBのチェックをするのではなく `assert_no_text` などをつかって表示がなくなったことを確認するようにしましょう。

yoshitsugu reviewed (commented):
コメントしました！

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#discussion_r1607385423
Systemテストとしててすとするために、assert_no_textを使用する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#discussion_r1607384008
モデルテストとして	言及機能のCRUDをテストした。

yoshitsugu commented on test/models/report_test.rb:
```diff
@@ -6,4 +6,54 @@ class ReportTest < ActiveSupport::TestCase
+    Report.create!(user: users(:alice), title: 'title', content: "http://localhost:3000/reports/#{reports(:report_alice).id}")
+    updated_report = Report.find_by(title: 'title')
+    updated_report.update!(title: 'title', content: "http://localhost:3000/reports/#{reports(:report_guest).id}")
+    assert_equal updated_report.id, reports(:report_guest).mentioned_reports.first.id
```
> 更新時に本文の内容に応じて、Mentionが追加、維持、削除されることまで確認しましょう。

と書きました。これだと「日報の更新時に新しいメンションが追加されていること」しかテストできていません。
「日報の更新時に本文中にそのままあるURLに対するメンションは維持されていること」「日報の更新時に本文中から削除されたメンションは削除されていること」も確認しましょう。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8#discussion_r1612497927
更新時に変更しないURLの言及は維持されること、または変更されて削除した言及は削除されていることをテストする。

yoshitsugu commented on test/models/report_test.rb:
```diff
@@ -56,4 +56,16 @@ class ReportTest < ActiveSupport::TestCase
     assert_equal updated_report.id, reports(:report_guest).mentioned_reports.first.id
   end
+  test '#update_mentions with multiple URLs' do
+    Report.create!(user: users(:alice), title: 'title', content: "http://localhost:3000/reports/#{reports(:report_alice).id}")
+    updated_report = Report.find_by(title: 'title')
```
```
report =  Report.create!(user: users(:alice), title: 'title', content: "http://localhost:3000/reports/#{reports(:report_alice).id}")
```
でよいのではないでしょうか。
titleって一意であることを保証されてないですよね？

yoshitsugu commented on test/models/report_test.rb:
```diff
@@ -56,4 +56,16 @@ class ReportTest < ActiveSupport::TestCase
+    assert_equal updated_report.id, reports(:report_guest).mentioned_reports.first.id
+  end
+  test '#update_mentions with deleted URL' do
+    Report.create!(user: users(:alice), title: 'title', content: "http://localhost:3000/reports/#{reports(:report_alice).id}")
+    updated_report = Report.find_by(title: 'title')
```
同様です。

yoshitsugu reviewed (commented):
コメントしました。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/8/files#r1615038826
report作成の手順を変更する。titleが一意であることが保証されてないため。

yoshitsugu reviewed (approved):
OKです！
